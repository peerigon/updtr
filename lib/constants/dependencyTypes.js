"use strict";

const REGULAR = 0;
const DEV = 1;
const OPTIONAL = 2;

function fromNpm(type) {
    switch (type) {
        case "dependencies":
            return REGULAR;
        case "devDependencies":
            return DEV;
        case "optionalDependencies":
            return OPTIONAL;
        default:
            throw new Error(`Unsupported dependency type: ${ type }`);
    }
}

exports.REGULAR = REGULAR;
exports.DEV = DEV;
exports.OPTIONAL = OPTIONAL;

exports.fromNpm = fromNpm;
// yarn uses the same strings. Change this if necessary.
exports.fromYarn = fromNpm;
