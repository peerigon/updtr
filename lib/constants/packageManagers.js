"use strict";

const NPM = "npm";
const YARN = "yarn";
const SUPPORTED = [NPM, YARN];

exports.NPM = NPM;
exports.YARN = YARN;
exports.SUPPORTED = SUPPORTED;
