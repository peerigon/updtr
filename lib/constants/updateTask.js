"use strict";

const WANTED = "wanted";
const LATEST = "latest";

exports.WANTED = WANTED;
exports.LATEST = LATEST;
