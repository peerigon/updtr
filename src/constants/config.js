// The first value in arrays is the default value

export const USE_NPM = "npm";
export const USE_YARN = "yarn";
export const USE_OPTIONS = [USE_NPM, USE_YARN];
export const UPDATE_TO_LATEST = "latest";
export const UPDATE_TO_NON_BREAKING = "non-breaking";
export const UPDATE_TO_WANTED = "wanted";
export const UPDATE_TO_OPTIONS = [
    UPDATE_TO_LATEST,
    UPDATE_TO_NON_BREAKING,
    UPDATE_TO_WANTED,
];
export const SAVE_SMART = "smart";
export const SAVE_CARET = "caret";
export const SAVE_EXACT = "exact";
export const SAVE_OPTIONS = [SAVE_SMART, SAVE_CARET, SAVE_EXACT];
