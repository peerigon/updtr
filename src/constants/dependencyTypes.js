export const REGULAR = "dependencies";
export const DEV = "devDependencies";
export const OPTIONAL = "optionalDependencies";

export function fromNpm(type) {
    switch (type) {
        case "dependencies":
            return REGULAR;
        case "devDependencies":
            return DEV;
        case "optionalDependencies":
            return OPTIONAL;
        default:
            throw new Error(`Unsupported dependency type: ${ type }`);
    }
}

export {
    // yarn uses the same strings. Change this if necessary.
    fromNpm as fromYarn,
};
