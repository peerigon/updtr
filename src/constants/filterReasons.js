export const GIT = "git";
export const EXOTIC = "exotic";
export const UNSTABLE = "unstable";
export const EXCLUDED = "excluded";
export const NOT_WANTED = "not-wanted";
