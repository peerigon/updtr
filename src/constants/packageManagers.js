export const NPM = "npm";
export const YARN = "yarn";
export const SUPPORTED = [NPM, YARN];
