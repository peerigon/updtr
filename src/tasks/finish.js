import Sequence from "./util/Sequence";
import { isUpdateToNonBreaking } from "./util/createUpdateTask";

function isIncompleteResult(result) {
    return result.success === true && isUpdateToNonBreaking(result) === true;
}

async function finishIncomplete(sequence, incomplete, allResults) {
    const updtr = sequence.updtr;
    const modulesToCheck = incomplete.map(result => result.name);
    const listCmd = updtr.cmds.list({ modules: modulesToCheck });

    let stdout;

    try {
        stdout = (await sequence.exec("list-incomplete", listCmd)).stdout;
    } catch (err) {
        // npm may exit with zero code 1 complaining about invalid installed versions
        if (err.code > 1) {
            throw err;
        }
        stdout = err.stdout;
    }
    const moduleVersions = updtr.parse.list(stdout, listCmd);

    return (
        allResults
            .map(result => {
                if (isIncompleteResult(result) === false) {
                    return result;
                }

                const version = moduleVersions.find(
                    module => module.name === result.name
                ).version;

                return {
                    ...result,
                    updateTo: version,
                };
            })
            // Remove results where no actual update did happen.
            // These results can happen if the updateTo option was set to non-breaking
            // and the module did not have a new version for the rollbackTo version range.
            .filter(result => result.rollbackTo !== result.updateTo)
    );
}

export default (async function finish(updtr, results) {
    const incomplete = results.filter(isIncompleteResult);
    const sequence = new Sequence("finish", updtr);
    let finishedResults = results;

    if (incomplete.length > 0) {
        sequence.start();
        sequence.emit("incomplete", { incomplete });
        finishedResults = await finishIncomplete(sequence, incomplete, results);
        sequence.end({ results: finishedResults });
    }

    return finishedResults;
});
