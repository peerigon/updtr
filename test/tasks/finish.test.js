import { USE_YARN } from "../../src/constants/config";
import finish from "../../src/tasks/finish";
import FakeUpdtr from "../helpers/FakeUpdtr";
import { npmList, yarnList } from "../fixtures/execResults";
import {
    module1ToLatestSuccess,
    module1ToNonBreakingSuccess,
    module2ToLatestSuccess,
    module2ToNonBreakingSuccess,
    module1ToLatestFail,
    module1ToNonBreakingFail,
    module2ToLatestFail,
    module2ToNonBreakingFail,
} from "../fixtures/updateResults";

describe("finish()", () => {
    describe("when the given results array is empty", () => {
        test("should resolve immediately with an empty array without emitting events", async () => {
            const updtr = new FakeUpdtr();

            expect(await finish(updtr, [])).toEqual([]);
            expect(updtr.exec.args).toEqual([]);
            expect(updtr.emit.args).toEqual([]);
        });
    });
    describe("when the given results are complete", () => {
        test("should resolve immediately with the given results without emitting events", async () => {
            const updtr = new FakeUpdtr();
            const results = [
                module1ToLatestSuccess,
                module1ToLatestFail,
                module1ToNonBreakingFail,
                module2ToLatestSuccess,
                module2ToLatestFail,
                module2ToNonBreakingFail,
            ];

            expect(await finish(updtr, results)).toEqual(results);
            expect(updtr.exec.args).toEqual([]);
            expect(updtr.emit.args).toEqual([]);
        });
    });
    describe("when there are incomplete results", () => {
        describe("using npm", () => {
            test("should return the expected results and emit the expected events", async () => {
                const updtr = new FakeUpdtr();
                const results = [
                    module1ToLatestSuccess,
                    module1ToNonBreakingSuccess,
                    module2ToLatestSuccess,
                    module2ToNonBreakingSuccess,
                ];

                updtr.execResults = npmList;

                expect(await finish(updtr, results)).toMatchSnapshot(
                    "incomplete > npm > results"
                );
                expect(updtr.exec.args).toMatchSnapshot(
                    "incomplete > npm > exec args"
                );
                expect(updtr.emit.args).toMatchSnapshot(
                    "incomplete > npm > emit args"
                );
            });
        });
        describe("using yarn", () => {
            test("should return the expected results and emit the expected events", async () => {
                const updtr = new FakeUpdtr({
                    use: USE_YARN,
                });
                const results = [
                    module1ToLatestSuccess,
                    module1ToNonBreakingSuccess,
                    module2ToLatestSuccess,
                    module2ToNonBreakingSuccess,
                ];

                updtr.execResults = yarnList;

                expect(await finish(updtr, results)).toMatchSnapshot(
                    "incomplete > yarn > results"
                );
                expect(updtr.exec.args).toMatchSnapshot(
                    "incomplete > yarn > exec args"
                );
                expect(updtr.emit.args).toMatchSnapshot(
                    "incomplete > yarn > emit args"
                );
            });
        });
    });

    //     return;
    //     describe("using npm", () => {
    //         describe("when the tests succeed", () => {
    //             test("should return the expected update results", async () => {
    //                 const updtr = new FakeUpdtr();
    //                 const updateTasks = createUpdateTasks(updtr.config);

    //                 updtr.execResults = update.concat(testPass);

    //                 const success = await batchUpdate(updtr, updateTasks);

    //                 expect(success).toBe(true);
    //             });
    //         });
    //         describe("when the test fails", () => {
    //             test("should emit expected events and execute expected commands", async () => {
    //                 const updtr = new FakeUpdtr();
    //                 const updateTasks = createUpdateTasks(updtr.config);

    //                 updtr.execResults = update.concat(testFail);

    //                 const success = await batchUpdate(updtr, updateTasks);

    //                 expect(updtr.exec.args).toMatchSnapshot();
    //                 expect(updtr.emit.args).toMatchSnapshot();
    //                 expect(success).toBe(false);
    //             });
    //         });
    //     });
    //     describe("using yarn", () => {
    //         describe("when the tests succeed", () => {
    //             test("should return the expected update results", async () => {
    //                 const updtr = new FakeUpdtr({
    //                     use: "yarn",
    //                 });
    //                 const updateTasks = createUpdateTasks(updtr.config);

    //                 updtr.execResults = update.concat(testPass);

    //                 const success = await batchUpdate(updtr, updateTasks);

    //                 expect(success).toBe(true);
    //             });
    //         });
    //         describe("when the test fails", () => {
    //             test("should emit expected events and execute expected commands", async () => {
    //                 const updtr = new FakeUpdtr({
    //                     use: "yarn",
    //                 });
    //                 const updateTasks = createUpdateTasks(updtr.config);

    //                 updtr.execResults = update.concat(testFail);

    //                 const success = await batchUpdate(updtr, updateTasks);

    //                 expect(updtr.exec.args).toMatchSnapshot();
    //                 expect(updtr.emit.args).toMatchSnapshot();
    //                 expect(success).toBe(false);
    //             });
    //         });
    //     });
    // });

    // describe("unexpected errors", () => {
    //     test("should completely bail out if the update cmd exits with a non-zero exit code", async () => {
    //         const updtr = new FakeUpdtr();
    //         const updateTasks = createUpdateTasks(updtr.config);
    //         let givenErr;

    //         updtr.execResults = errorExecUpdate;

    //         try {
    //             await batchUpdate(updtr, updateTasks);
    //         } catch (err) {
    //             givenErr = err;
    //         }

    //         expect(givenErr).toBe(execError);
    //         // emitted events: start, updating
    //         expect(updtr.emit.args.length).toBe(2);
    //     });
    //     test("should completely bail out if the rollback cmd exits with a non-zero exit code", async () => {
    //         const updtr = new FakeUpdtr();
    //         const updateTasks = createUpdateTasks(updtr.config);
    //         let givenErr;

    //         updtr.execResults = errorExecRollback;

    //         try {
    //             await batchUpdate(updtr, updateTasks);
    //         } catch (err) {
    //             givenErr = err;
    //         }

    //         expect(givenErr).toBe(execError);
    //         // emitted events: start, updating, testing, testResult, rollback
    //         expect(updtr.emit.args.length).toBe(5);
    //     });
    // });
});
