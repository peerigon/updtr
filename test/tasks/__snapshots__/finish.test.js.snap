// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`incomplete > npm > emit args 1`] = `
Array [
  Array [
    "finish/start",
    Object {},
  ],
  Array [
    "finish/incomplete",
    Object {
      "incomplete": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "success": true,
          "updateTo": "^1.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "success": true,
          "updateTo": "^2.0.0",
        },
      ],
    },
  ],
  Array [
    "finish/list-incomplete",
    Object {
      "cmd": "npm ls --json --depth=0 updtr-test-module-1 updtr-test-module-2",
    },
  ],
  Array [
    "finish/end",
    Object {
      "results": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "success": true,
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "success": true,
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "success": true,
          "updateTo": "2.1.1",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "success": true,
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
]
`;

exports[`incomplete > npm > exec args 1`] = `
Array [
  Array [
    "npm ls --json --depth=0 updtr-test-module-1 updtr-test-module-2",
  ],
]
`;

exports[`incomplete > npm > results 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": true,
    "updateTo": "2.0.0",
  },
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": true,
    "updateTo": "2.0.0",
  },
  Object {
    "name": "updtr-test-module-2",
    "rollbackTo": "2.0.0",
    "success": true,
    "updateTo": "2.1.1",
  },
  Object {
    "name": "updtr-test-module-2",
    "rollbackTo": "2.0.0",
    "success": true,
    "updateTo": "2.1.1",
  },
]
`;

exports[`incomplete > yarn > emit args 1`] = `
Array [
  Array [
    "finish/start",
    Object {},
  ],
  Array [
    "finish/incomplete",
    Object {
      "incomplete": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "success": true,
          "updateTo": "^1.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "success": true,
          "updateTo": "^2.0.0",
        },
      ],
    },
  ],
  Array [
    "finish/list-incomplete",
    Object {
      "cmd": "npm ls --json --depth=0 updtr-test-module-1 updtr-test-module-2",
    },
  ],
  Array [
    "finish/end",
    Object {
      "results": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "success": true,
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "success": true,
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "success": true,
          "updateTo": "2.1.1",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "success": true,
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
]
`;

exports[`incomplete > yarn > exec args 1`] = `
Array [
  Array [
    "npm ls --json --depth=0 updtr-test-module-1 updtr-test-module-2",
  ],
]
`;

exports[`incomplete > yarn > results 1`] = `
Array [
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": true,
    "updateTo": "2.0.0",
  },
  Object {
    "name": "updtr-test-module-1",
    "rollbackTo": "1.0.0",
    "success": true,
    "updateTo": "2.0.0",
  },
  Object {
    "name": "updtr-test-module-2",
    "rollbackTo": "2.0.0",
    "success": true,
    "updateTo": "2.1.1",
  },
  Object {
    "name": "updtr-test-module-2",
    "rollbackTo": "2.0.0",
    "success": true,
    "updateTo": "2.1.1",
  },
]
`;
